validation/unique_test.go

```diff
@@ -370,6 +370,7 @@ func TestUniqueArrayValidator(t *testing.T) {
 			{dialect: "mysql", expected: "WITH ctx_values(id, i) AS (SELECT * FROM (VALUES ROW(2,0),ROW(7,1),ROW(6,2)) t) SELECT i FROM ctx_values LEFT JOIN `models` ON `models`.`name` = ctx_values.id WHERE `models`.`name` IS NOT NULL"},
 			{dialect: "postgres", expected: "WITH ctx_values(id, i) AS (SELECT * FROM (VALUES (2,0::int),(7,1::int),(6,2::int)) t) SELECT i FROM ctx_values LEFT JOIN `models` ON `models`.`name` = ctx_values.id WHERE `models`.`name` IS NOT NULL"},
 			{dialect: "mssql", expected: "WITH ctx_values(id, i) AS (SELECT * FROM (VALUES (2,0),(7,1),(6,2)) t(id,i)) SELECT i FROM ctx_values LEFT JOIN `models` ON `models`.`name` = ctx_values.id WHERE `models`.`name` IS NOT NULL"},
+			{dialect: "clickhouse", expected: "WITH ctx_values(id, i) AS (SELECT * FROM (VALUES 'a UInt64, b UInt64', (2,0),(7,1),(6,2)) t) SELECT i FROM ctx_values LEFT JOIN `models` ON `models`.`name` = ctx_values.id WHERE `models`.`name` IS NOT NULL"},
```

Comment:
I tested the query on a running Clickhouse docker image and it doesn't work. The following query somewhat works though:
```sql
WITH ctx_values AS
    (
        SELECT *
        FROM VALUES('id UInt64, i UInt64', (2, 0), (7, 1), (6, 2))
    )
SELECT i
FROM ctx_values
LEFT JOIN users ON users.id = ctx_values.id
WHERE users.id IS NOT NULL
```
The query has a valid syntax, however it doesn't work as we want it to work. In this engine, it looks like the type's **zero value** is returned instead of `NULL` when the join doesn't match the `ON` condition. I tested it like so:
```
WITH ctx_values AS
    (
        SELECT *
        FROM VALUES('id UInt64, i UInt64', (1, 0), (2, 1), (4, 2))
    )
SELECT
    id,
    i,
    users.id,
    users.email
FROM ctx_values
LEFT JOIN users ON users.id = ctx_values.id
WHERE users.id IS NOT NULL

Query id: 4ab27032-b547-4b22-9427-6c63f8d0fb61

   ┌─id─┬─i─┬─users.id─┬─email─┐
1. │  1 │ 0 │        1 │ k     │
2. │  2 │ 1 │        0 │       │
3. │  4 │ 2 │        0 │       │
   └────┴───┴──────────┴───────┘

3 rows in set. Elapsed: 0.002 sec. 
```
This query should only return the first row but it doesn't because users.id defaults to 0 instead of NULL.

---

It's getting very different from the rest of the database engine so we probably should create a separate function to avoid overcomplicating `buildQuery()`:

```go
func (v *ExistsArrayValidator[T]) buildQuery(values []T, condition bool) *gorm.DB {
	dbType := v.Config().GetString("database.connection")
	if dbType == "clickhouse" {
		return v.buildClickhouseQuery(values, condition)
	}
        //...
```

Another important thing: the `id` column in the `ctx_values`  won't always be `uint64`. It will be of type `T`, the generic parameter of `ExistsArrayValidator`. I'm not sure how that works in ClickHouse. Maybe there's a way to get the clickhouse type name from a Go type in the [official lib](https://github.com/ClickHouse/clickhouse-go)? 

This really isn't very practical...

Reply:
Thanks for this investigation, I see what you mean and had been wondering about this myself. 

Regarding the UInt64 type defaulting to null, it looks like we would get the desired behavior from an `INNER JOIN` rather than a `LEFT JOIN`: 

`WITH ctx_values AS
    (
        SELECT *
        FROM VALUES('id UInt64, i UInt64', (1, 0), (2, 1), (4, 2))
    )
SELECT
    id,
    i,
    test.users.id,
    test.users.email
FROM ctx_values
INNER JOIN test.users ON test.users.id = ctx_values.id
WHERE test.users.id IS NOT NULL

Query id: 2e169b8f-6fcf-4e7a-9fcd-689e5ff42031

   ┌─id─┬─i─┬─users.id─┬─email─┐
1. │  1 │ 0 │        1 │ k     │
   └────┴───┴──────────┴───────┘

1 row in set. Elapsed: 0.010 sec.`

This could also be accomplished by a `LEFT JOIN` when the JOIN setting "[join_use_nulls](https://clickhouse.com/docs/en/operations/settings/settings#join_use_nulls)" is set to 1: 

`SET join_use_nulls = 1

Query id: 3a88865e-25fe-4ade-aeea-87f24c956eee

Ok.

0 rows in set. Elapsed: 0.004 sec.

...

WITH ctx_values AS
    (
        SELECT *
        FROM VALUES('id UInt64, i UInt64', (1, 0), (2, 1), (4, 2))
    )
SELECT
    id,
    i,
    test.users.id,
    test.users.email
FROM ctx_values
LEFT JOIN test.users ON test.users.id = ctx_values.id
WHERE test.users.id IS NOT NULL

Query id: d55eadea-2bcb-455b-873a-77205f4d0bfc

   ┌─id─┬─i─┬─users.id─┬─email─┐
1. │  1 │ 0 │        1 │ k     │
   └────┴───┴──────────┴───────┘

1 row in set. Elapsed: 0.009 sec.`

Do you have any opinions on how the best way to deal with this? 

I will keep looking into the existence of a reliable Go type <> Clickhouse type mapping.

Reply:
Ah, looks like this might be what we needed: https://github.com/ClickHouse/clickhouse-go/blob/main/TYPES.md